feat(op-node/op-batcher/op-proposer): add L1 fallbackClient

welkin22 opened the pull request:
**Description**

The FallbackClient is designed to automatically switch to an alternative L1 endpoint when the current one encounters issues, and revert back once the primary endpoint is functional again.It can reduce the instability of the L2 chain caused by L1 endpoint instability.

The core logic operates as follows: When an error occurs, it is recorded until either the error count surpasses a predetermined threshold or the count is reset by a timer that executes once per minute. If the threshold is exceeded within a one-minute span, it suggests the L1 endpoint has become unreliable, prompting the system to select the next address in the L1 URL. A new RPC client is then created to replace the current one. Simultaneously, a separate Goroutine monitors the initial endpoint's health and, when deemed stable, reverts back to its use.

As the clients employed in op-node and op-batcher/op-proposer differ significantly, two distinct fallbackClients have been implemented. The op-node version incorporates additional features such as subscription management and RPC legality checks, resulting in a more complex implementation.

Example of l1 flag after code modification:
```bash
--l1=https://addr1.com:8545,https://addr2.com:8545,https://addr3.com:8545
```

**Tests**
I have tested it in the local environment. I am going to add an e2e case, but I encountered some problems when executing the  `make devnet-allocs` command.I'm still trying to solve it

welkin22 committed:
feat(op-node/op-batcher/op-proposer): add fallbackClient

tynes commented on op-service/client/dial.go:
```diff
@@ -62,3 +66,50 @@ func dialRPCClientWithBackoff(ctx context.Context, log log.Logger, addr string)
+	return DialEthClientWithTimeout(timeout, l, url)
+}
+
+type EthClient interface {
```
Doesn't this interface already exist in go-ethereum?

welkin22 replied:
Hi, I couldn't find this interface in go-ethereum. I noticed that there is an ETHBackend interface in op-service/txmgr/txmgr.go:63. Also notice that op-node/client/client.go:15 has a Client interface. Neither of these interfaces can be used directly, so I imitated them and wrote my own interface. Could you please tell me the location of the interface in go-ethereum?

tynes replied:
Are you able to use these interfaces? https://github.com/ethereum/go-ethereum/blob/545f4c5547178bc8bde6af08b3ccaf68ca27f2c0/interfaces.go

welkin22 replied:
I can't use this because its interface is too fragmented. For example, in the batcher, both `HeaderByNumber` and `SendTransaction` are required, but they come from two different interfaces. I cannot return multiple interfaces in `DialEthClientWithTimeoutAndFallback`. I need to return a unified EthClient.

tynes commented:
This seems like its more lines of code than necessary for this kind of functionality, there are also very few comments. Perhaps aligning on the desire to have a fallback RPC client first is a better way to approach opening pull requests

trianglesphere commented on op-batcher/flags/flags.go:
```diff
@@ -26,7 +26,7 @@ var (
 	// Required flags
 	L1EthRpcFlag = &cli.StringFlag{
```
A string slice flag would be better than manually parsing strings.

welkin22 replied:
Yes I will try to modify this

welkin22 replied:
resolved

trianglesphere commented on op-node/sources/fallback_client.go:
```diff
@@ -0,0 +1,280 @@
+// FallbackClient is an RPC client, it can automatically switch to the next l1 endpoint
+// when there is a problem with the current l1 endpoint
+// and automatically switch back after the first l1 endpoint recovers.
+type FallbackClient struct {
```
This may be conflicting with https://github.com/ethereum-optimism/optimism/pull/7029, but the fallback client should be layered onto the client & be easy to create.

welkin22 replied:
It seems that #7029 was not merged, but rather closed. Right now, I can't be sure what it will turn out to be like.

> the fallback client should be layered onto the client & be easy to create.

I tried to achieve this goal, but since the fallback client needs to create multiple underlying RPC clients instead of just adding new functions to a single client, the situation is a bit more complicated. Currently I try to implement the client.RPC interface and wrap the original client.NewRPC when multiple addresses are configured. See [fallbackClientWrap ](https://github.com/ethereum-optimism/optimism/pull/7311/files#diff-a27156a67289c4da44b8746353b897cde1e4fb77737af85838ad9ca91923ef12R197) method.
This is the easiest way I can think of to create it. If you have any better ideas, please let me know.

trianglesphere reviewed (commented):
I understand the desire for this feature, but I am incredibly worried about consistency between different RPCs. Different RPC providers will not be in sync & these components are designed to handle re-orgs, but may not handle data inconsistencies other than re-orgs.

In addition, when a re-org is detected, a significant amount of work has to be done in the op-batcher/op-node so it is important than this client does not show something like a re-org occuring.

welkin22 commented:
> This seems like its more lines of code than necessary for this kind of functionality, there are also very few comments. Perhaps aligning on the desire to have a fallback RPC client first is a better way to approach opening pull requests

@tynes Hi, I believe that all the code is meant to implement a fallback client. The code appears to have more lines than anticipated because it attempts to merge the fallback client code with the existing code. For example, the places where *ethclient.Client is referenced are changed to ethereum.ChainStateReader, as only by replacing it with an interface can different implementations be introduced.
Sorry about the few comments, I will add more comments to help understand the code.

welkin22 commented:
> I understand the desire for this feature, but I am incredibly worried about consistency between different RPCs. Different RPC providers will not be in sync & these components are designed to handle re-orgs, but may not handle data inconsistencies other than re-orgs.
> 
> In addition, when a re-org is detected, a significant amount of work has to be done in the op-batcher/op-node so it is important than this client does not show something like a re-org occuring.

@trianglesphere Hi, I understand your concerns and I also considered this issue at the beginning of the design. Perhaps we can expand on this concern and discuss it.First of all, the l1 endpoint we fill in is most of the time a node we trust. The inconsistency in its data will only be the difference in the latest block height, and other inconsistencies rarely occur. Secondly, I have not found any examples in the code of op-node and op-batcher where data inconsistency may cause problems. It seems that there are necessary checking and retry mechanisms everywhere. Compared with the serious situation where the endpoint of l1 is paralyzed and cannot be accessed, causing the block to be unable to be produced, triggering the re-org logic and retry mechanism seems to be worthwhile.
If you could further point out a specific data inconsistency problem case, I would be very grateful.

tynes commented:
> > This seems like its more lines of code than necessary for this kind of functionality, there are also very few comments. Perhaps aligning on the desire to have a fallback RPC client first is a better way to approach opening pull requests
> 
> @tynes Hi, I believe that all the code is meant to implement a fallback client. The code appears to have more lines than anticipated because it attempts to merge the fallback client code with the existing code. For example, the places where *ethclient.Client is referenced are changed to ethereum.ChainStateReader, as only by replacing it with an interface can different implementations be introduced. Sorry about the few comments, I will add more comments to help understand the code.

What I mean here is that I believe your implementation needs to be simplified before considered for merge. For example, there are many fields that can be removed from this struct

```go
type FallbackClient struct {
	firstRpc          client.RPC
	urlList           []string
	rpcInitFunc       func(url string) (client.RPC, error)
	lastMinuteFail    atomic.Int64
	currentRpc        atomic.Pointer[client.RPC]
	currentIndex      int
	mx                sync.Mutex
	log               log.Logger
	isInFallbackState bool
	subscribeFunc     func() (event.Subscription, error)
	l1HeadsSub        *ethereum.Subscription
	l1ChainId         *big.Int
	l1Block           eth.BlockID
	ctx               context.Context
	isClose           chan struct{}
	metrics           metrics.Metricer
}
```

Why does it need `firstRPC` and `currentRPC`? Why does it need `isInFallbackState`? Why does it have a `l1ChainId` and a `l1Block`? Having all of these makes the internals more complex than it needs to be.

I have a preference for the functionality to be as follows:
- upon construction, make an ethclient for each url
- when caller calls the client:
  - for each client, submit the request
  - compare the results
  - return error if mismatch
  - ignore backend errors/timeouts

This ensures that the backends are consistent altho it does use a linear number of requests. This feature is more useful for power users and power users can run their own nodes. What do you think @trianglesphere ?

tynes commented:
> For example, the places where *ethclient.Client is referenced are changed to ethereum.ChainStateReader, as only by replacing it with an interface can different implementations be introduced.

I am supportive of this kind of change and it is something that could easily be broken out into its own PR, making things overall easier to review/merge

welkin22 committed:
l1 flag use StringSliceFlag

tynes commented on op-node/sources/fallback_client.go:
```diff
@@ -0,0 +1,280 @@
+	"time"
+)
+
+// FallbackClient is an RPC client, it can automatically switch to the next l1 endpoint
```
nit: it shouldn't care if its L1 or not L1

welkin22 replied:
removed

tynes commented on op-node/sources/fallback_client.go:
```diff
@@ -0,0 +1,280 @@
+
+// NewFallbackClient returns a new FallbackClient. l1ChainId and l1Block are used to check
+// whether the newly switched rpc is legal.
+func NewFallbackClient(ctx context.Context, rpc client.RPC, urlList []string, log log.Logger, l1ChainId *big.Int, l1Block eth.BlockID, rpcInitFunc func(url string) (client.RPC, error)) client.RPC {
```
Do we need 2 implementations or is it possible to use a single impl from `op-service`?

welkin22 replied:
Tried it before but had circular dependency issues here, I'll try it again

tynes commented on op-node/sources/fallback_client.go:
```diff
@@ -0,0 +1,280 @@
+		l.log.Info("recover the current rpc to the first rpc", "url", l.urlList[0])
+	}()
+}
+
```
Can you add godocs to the functions?

welkin22 replied:
ok, I would add them

welkin22 commented:
> I have a preference for the functionality to be as follows:
> 
> * upon construction, make an ethclient for each url
> * when caller calls the client:
>   
>   * for each client, submit the request
>   * compare the results
>   * return error if mismatch
>   * ignore backend errors/timeouts
> 
> This ensures that the backends are consistent altho it does use a linear number of requests. This feature is more useful for power users and power users can run their own nodes. What do you think @trianglesphere ?

@tynes Hi, I thought about the method you mentioned. It is indeed simple enough, but there are some problems that need to be solved:
1. It is very difficult to compare the return value of rpc. After all, result is an any type. Is there any way to quickly compare the return values? Will the process of comparing return values slow down the overall request speed?
2. Multiple endpoints are requested at the same time. If the performance of one of them is poor (but less than the timeout period), it will affect the performance of the entire request.
3. If all endpoints fail, but the returned errors are different, which one should I use?
If all of the above issues can be resolved, I would be willing to re-implement a new version following this approach.

> Why does it need `firstRPC` and `currentRPC`? Why does it need `isInFallbackState`? Why does it have a `l1ChainId` and a `l1Block`? Having all of these makes the internals more complex than it needs to be.
> 
firstRPC is the rpc that is used daily. Both `currentRPC` and `isInFallbackState` are used to manage the state when temporarily switching to an alternative rpc. The current design idea is that firstRPC will not fail all the time, but will only fail temporarily for a period of time. It usually works normally. We only need to switch to the alternative RPC during the failure. When firstRPC recovers, it will automatically switch back.
a `l1ChainId` and a `l1Block` are used for checking the L1 endpoint. I noticed that `ValidateL1Config` has this logic, so when switching to a new RPC, this check should also be performed.

welkin22 committed:
remove l1 from fallback client

welkin22 commented:
> > For example, the places where *ethclient.Client is referenced are changed to ethereum.ChainStateReader, as only by replacing it with an interface can different implementations be introduced.
> 
> I am supportive of this kind of change and it is something that could easily be broken out into its own PR, making things overall easier to review/merge

Indeed. I would try splitting into two PRs, one containing the fallback client itself, and one trying to integrate the fallback client into other code

tynes commented:
> > > For example, the places where *ethclient.Client is referenced are changed to ethereum.ChainStateReader, as only by replacing it with an interface can different implementations be introduced.
> > 
> > 
> > I am supportive of this kind of change and it is something that could easily be broken out into its own PR, making things overall easier to review/merge
> 
> Indeed. I would try splitting into two PRs, one containing the fallback client itself, and one trying to integrate the fallback client into other code

Would very much appreciate this approach :) thank you!

For comparing type `any`, shouldn't `reflect.DeepEqual` work? Also I was thinking of the `ethclient.Client` interface but with fallback support under the hood. https://github.com/ethereum/go-ethereum/blob/master/ethclient/ethclient.go

tynes commented:
> > I have a preference for the functionality to be as follows:
> > 
> > * upon construction, make an ethclient for each url
> > * when caller calls the client:
> >   
> >   * for each client, submit the request
> >   * compare the results
> >   * return error if mismatch
> >   * ignore backend errors/timeouts
> > 
> > This ensures that the backends are consistent altho it does use a linear number of requests. This feature is more useful for power users and power users can run their own nodes. What do you think @trianglesphere ?
> 
> @tynes Hi, I thought about the method you mentioned. It is indeed simple enough, but there are some problems that need to be solved:
> 
> 1. It is very difficult to compare the return value of rpc. After all, result is an any type. Is there any way to quickly compare the return values? Will the process of comparing return values slow down the overall request speed?
> 2. Multiple endpoints are requested at the same time. If the performance of one of them is poor (but less than the timeout period), it will affect the performance of the entire request.
> 3. If all endpoints fail, but the returned errors are different, which one should I use?
>    If all of the above issues can be resolved, I would be willing to re-implement a new version following this approach.
> 
> > Why does it need `firstRPC` and `currentRPC`? Why does it need `isInFallbackState`? Why does it have a `l1ChainId` and a `l1Block`? Having all of these makes the internals more complex than it needs to be.
> 
> firstRPC is the rpc that is used daily. Both `currentRPC` and `isInFallbackState` are used to manage the state when temporarily switching to an alternative rpc. The current design idea is that firstRPC will not fail all the time, but will only fail temporarily for a period of time. It usually works normally. We only need to switch to the alternative RPC during the failure. When firstRPC recovers, it will automatically switch back. a `l1ChainId` and a `l1Block` are used for checking the L1 endpoint. I noticed that `ValidateL1Config` has this logic, so when switching to a new RPC, this check should also be performed.

I would like to get some more feedback from my peers as well because this is a big design space, will follow up

welkin22 committed:
Delete other code and leave only fallback client

welkin22 committed:
add comment to op-node's FallbackClient

welkin22 commented:
> > > > For example, the places where *ethclient.Client is referenced are changed to ethereum.ChainStateReader, as only by replacing it with an interface can different implementations be introduced.
> > > 
> > > 
> > > I am supportive of this kind of change and it is something that could easily be broken out into its own PR, making things overall easier to review/merge
> > 
> > 
> > Indeed. I would try splitting into two PRs, one containing the fallback client itself, and one trying to integrate the fallback client into other code
> 
> Would very much appreciate this approach :) thank you!
> 
> For comparing type `any`, shouldn't `reflect.DeepEqual` work? Also I was thinking of the `ethclient.Client` interface but with fallback support under the hood. https://github.com/ethereum/go-ethereum/blob/master/ethclient/ethclient.go

@tynes Hi, I have deleted the code except fallbackClient and added some comments. Now we can focus on the discussion of fallbackClient implementation.
One problem is that I still can't write fallbackClient only in op-service. This is mainly because it is difficult to find a unified interface entry point. In op-node, if I want to replace the implementation, I can only start with the client.RPC interface. Everything else is a structure. If the implementation of the structure is replaced, all references to the structure must be modified. The Client in ethclient.go you mentioned is still a structure, and its implementation cannot be easily replaced. The same situation exists in op-batcher and op-proposer, so I chose to define the EthClient interface myself. But if I want to use this interface in op-node, I still need to modify a lot of code.

welkin22 commented:
> > I understand the desire for this feature, but I am incredibly worried about consistency between different RPCs. Different RPC providers will not be in sync & these components are designed to handle re-orgs, but may not handle data inconsistencies other than re-orgs.
> > In addition, when a re-org is detected, a significant amount of work has to be done in the op-batcher/op-node so it is important than this client does not show something like a re-org occuring.
> 
> @trianglesphere Hi, I understand your concerns and I also considered this issue at the beginning of the design. Perhaps we can expand on this concern and discuss it.First of all, the l1 endpoint we fill in is most of the time a node we trust. The inconsistency in its data will only be the difference in the latest block height, and other inconsistencies rarely occur. Secondly, I have not found any examples in the code of op-node and op-batcher where data inconsistency may cause problems. It seems that there are necessary checking and retry mechanisms everywhere. Compared with the serious situation where the endpoint of l1 is paralyzed and cannot be accessed, causing the block to be unable to be produced, triggering the re-org logic and retry mechanism seems to be worthwhile. If you could further point out a specific data inconsistency problem case, I would be very grateful.

Sorry, I found that it is still necessary to avoid re-org . I will try to let the fallback client avoid re-org .

welkin22 committed:
add comment to op-batcher's FallbackClient

sebastianst commented:
Converted to draft until we have decided on how to proceed and tests are added.
